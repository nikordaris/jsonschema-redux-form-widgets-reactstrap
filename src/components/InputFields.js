// @flow
import createInputField from '../createInputField';

export const Text = createInputField({ type: 'text' });
export const Email = createInputField({ type: 'email' });
export const Password = createInputField({ type: 'password' });
export const Date = createInputField({ type: 'date' });
export const DateTime = createInputField({ type: 'datetime-local' });
export const Number = createInputField({
  type: 'number',
  parse: value => {
    try {
      return parseInt(value)
    }
    catch (Exception) {
      return value;
    }
  }
});
export const Color = createInputField({
  type: 'color',
  styles: { input: { height: '40px' } }
});
export const TextArea = createInputField({ type: 'textarea' });
