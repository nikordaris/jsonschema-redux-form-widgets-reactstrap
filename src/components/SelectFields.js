// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Field, change, untouch } from 'redux-form';
import { Input, Label } from 'reactstrap';
import { sortBy, omit, isEmpty, forEach, get, has } from 'lodash';
import SchemaVis, { hasComponent } from 'react-jsonschema-vis';

import FormField from './FormField';
import { injectSheet } from '../Jss';
import validate from '../validator';

@injectSheet({
  selectedObject: { marginTop: 10 },
  card: { marginBottom: 10, marginTop: 15 },
  header: { padding: 5, paddingLeft: 10 }
})
@connect(
  () => ({}),
  dispatch => bindActionCreators({ change, untouch }, dispatch),
  undefined,
  { withRef: true }
)
export class SingleSelect extends Component {
  static defaultProps = {
    initialSelected: '',
    onChange: () => { },
    change: () => { },
    untouch: () => { }
  };

  state = {
    selected: '',
    showLabel: true
  };

  props: {
    schemaVis: SchemaVisType,
    initialSelected?: string,
    form: string,
    name: string,
    required: boolean,
    classes: { [string]: any },
    sheet: any,
    styles: { [string]: any },
    showLabel: boolean,
    getOptions: (schema: { [string]: any }) => Array<{ label: string, value: any, tooltip: string }>,
    change: (form: string, prop: string, value: any) => void,
    untouch: (form: string, fields: string) => void,
    onChange: (selected: string) => void
  };

  state: {
    selected?: string
  };

  componentWillReceiveProps(nextProps: any) {
    const { initialSelected: prevInitialSelected } = this.props;
    const { initialSelected } = nextProps;
    if (initialSelected !== prevInitialSelected) {
      this.setState({ ...this.state, selected: initialSelected });
    }
  }

  componentWillMount() {
    const { initialSelected } = this.props;
    if (initialSelected) {
      this.setState({ ...this.state, selected: initialSelected });
    }
  }

  getOptions = (schema: { [string]: any }) => {
    const {
      name,
      required,
      classes,
      sheet,
      styles,
      change,
      untouch,
      onChange,
      schemaVis,
      schemaVis: { prefix },
      showLabel,
      initialSelected,
      ...rest
    } = this.props;
    const [schemas, index] = has(schema, 'oneOf')
      ? [schema.oneOf, `${name}-oneOf`]
      : [schema.anyOf, `${name}-anyOf`];
    return schemas
      .map((schema, idx) => {
        const { id, title, const: value, description } = schema;
        const rendered =
          (hasComponent(schema, prefix) || !isEmpty(schema.properties)) &&
          <SchemaVis
            key={`${index}-${idx}`}
            {...schemaVis}
            {...rest}
            namespace={name}
            schema={schema}
          />;

        return {
          label: title || id || value,
          tooltip: description,
          value: rendered || value || id || title
        };
      })
      .filter(o => o.value);
  };

  renderInputOptions(options: Array<OptionType>) {
    return sortBy(options, o => o.label).map(({ label, value }, idx) => (
      <option key={idx} value={React.isValidElement(value) ? label : value}>
        {label}
      </option>
    ));
  }

  handleChange = (e: { target: { value: string } }) => {
    const {
      name,
      change,
      untouch,
      form,
      onChange,
      schemaVis: { schema }
    } = this.props;
    const selected = e.target.value;
    const items = get(schema, 'oneOf', schema.anyOf);
    if (name) {
      change(form, name, {});
    }
    items.forEach(s =>
      forEach(s.properties, (_, p) => untouch(form, name ? `${name}.${p}` : p))
    );
    this.setState({ ...this.state, selected });
    onChange(selected);
  };

  renderSelectInput(options: Array<OptionType>) {
    const { showLabel, schemaVis: { schema }, classes } = this.props;
    return (
      <div>
        {showLabel && <Label>{schema.title}</Label>}
        <Input
          className={classes.select}
          type="select"
          onChange={this.handleChange}
          value={this.state.selected}
        >
          <option disabled value="">Select {schema.title}</option>
          {this.renderInputOptions(options)}
        </Input>
        {this.renderInputObject(options)}
      </div>
    );
  }

  renderSelectField(options: Array<OptionType>) {
    const rest = omit(this.props, [
      'tag',
      'children',
      'change',
      'untouch',
      'schemaVis',
      'styles',
      'onChange',
      'showLabel',
      'initialSelected'
    ]);
    const { schemaVis: { schema }, required } = this.props;
    return (
      <Field
        type="select"
        validate={validate(schema, required)}
        component={FormField}
        schema={schema}
        {...rest}
      >
        <Input type="select">
          <option disabled value="">Select {schema.title}</option>
          {this.renderInputOptions(options)}
        </Input>
      </Field>
    );
  }

  renderInputObject(options: Array<OptionType>) {
    const { classes } = this.props;
    const { selected } = this.state;

    const selectedObj = selected && options.find(o => o.label === selected);
    if (selectedObj && selectedObj.value) {
      return (
        <div className={classes.selectedObject}>
          {selectedObj.value}
        </div>
      );
    }

    return undefined;
  }

  render() {
    const { getOptions = this.getOptions, schemaVis: { schema } } = this.props;

    const options = getOptions(schema);
    if (isEmpty(options.filter(({ value }) => React.isValidElement(value)))) {
      return this.renderSelectField(options);
    }

    return this.renderSelectInput(options);
  }
}
