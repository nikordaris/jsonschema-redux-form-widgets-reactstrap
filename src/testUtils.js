import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { reduxForm, reducer as formReducer } from 'redux-form';
import { createStore, combineReducers } from 'redux';
import { shallow } from 'enzyme';

export const matchSnapshotWithReduxForm = (options, TestComponent) => () => {
  const rootReducers = combineReducers({ form: formReducer });
  const store = createStore(rootReducers);
  const WrappedComponent = reduxForm({ form: 'MyForm' })(TestComponent);
  const tree = renderer
    .create(
      <Provider store={store}>
        <WrappedComponent {...options} />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
};

export const matchSnapshot = (options, TestComponent) => () => {
  const tree = renderer.create(<TestComponent {...options} />).toJSON();
  expect(tree).toMatchSnapshot();
};

export const matchSnapshotShallow = (options, TestComponent) => () => {
  const wrapper = shallow(<TestComponent {...options} />);

  expect(wrapper.getNodes()).toMatchSnapshot();
};
