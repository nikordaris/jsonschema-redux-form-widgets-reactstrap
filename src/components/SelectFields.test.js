import React from 'react';
import { Input } from 'reactstrap';
import renderer from 'react-test-renderer';
import { SingleSelect } from './SelectFields';
import { Provider } from 'react-redux';
import { reduxForm, reducer as formReducer } from 'redux-form';
import { createStore, combineReducers } from 'redux';
import { mount } from 'enzyme';
import { Text } from './InputFields';
import { matchSnapshotWithReduxForm } from '../testUtils';

describe('Render SelectFields', () => {
  it(
    'Should render SingleSelect',
    matchSnapshotWithReduxForm(
      {
        schemaVis: {
          prefix: 'meta.vis',
          meta: {
            hasComponent: () => false
          },
          schema: {
            id: 'SingleSelect',
            title: 'SingleSelect',
            type: 'number',
            oneOf: [
              { title: 'foo', const: 1, description: 'foo description' },
              { title: 'bar', const: 2, description: 'bar description' }
            ]
          }
        },
        name: 'SingleSelect'
      },
      SingleSelect
    )
  );

  it(
    'Should render SingleSelect with only values',
    matchSnapshotWithReduxForm(
      {
        schemaVis: {
          prefix: 'meta.vis',
          meta: {
            hasComponent: () => false
          },
          schema: {
            id: 'SingleSelect',
            title: 'SingleSelect',
            type: 'string',
            oneOf: [{ const: 'foo' }, { const: 'bar' }]
          }
        },
        name: 'SingleSelect'
      },
      SingleSelect
    )
  );

  it(
    'Should render SingleSelect with only titles',
    matchSnapshotWithReduxForm(
      {
        schemaVis: {
          prefix: 'meta.vis',
          meta: {
            hasComponent: () => false
          },
          schema: {
            id: 'SingleSelect',
            title: 'SingleSelect',
            type: 'string',
            oneOf: [{ title: 'foo' }, { title: 'bar' }]
          }
        },
        name: 'SingleSelect'
      },
      SingleSelect
    )
  );

  it(
    'Should render object SingleSelect',
    matchSnapshotWithReduxForm(
      {
        schemaVis: {
          components: { Text },
          prefix: 'meta.vis',
          meta: {
            hasComponent: () => true
          },
          schema: {
            id: 'SingleSelect',
            title: 'SingleSelect',
            type: 'object',
            oneOf: [
              {
                title: 'Foo',
                id: 'Foo',
                type: 'object',
                properties: {
                  x: {
                    type: 'string',
                    title: 'X',
                    id: 'x',
                    meta: {
                      vis: {
                        component: 'Text'
                      }
                    }
                  }
                }
              },
              {
                id: 'Bar',
                title: 'Bar',
                type: 'object',
                properties: {
                  y: {
                    type: 'string',
                    title: 'Y',
                    id: 'Y',
                    meta: {
                      vis: {
                        component: 'Text'
                      }
                    }
                  }
                }
              }
            ]
          }
        },
        name: 'SingleSelect'
      },
      SingleSelect
    )
  );

  it('should handle select change', () => {
    const rootReducers = combineReducers({ form: formReducer });
    const store = createStore(rootReducers);
    const WrappedComponent = reduxForm({ form: 'MyForm' })(SingleSelect);
    const schema = {
      id: 'SingleSelect',
      title: 'SingleSelect',
      type: 'object',
      oneOf: [
        {
          title: 'Foo',
          id: 'Foo',
          type: 'object',
          properties: {
            x: {
              type: 'string',
              title: 'X',
              id: 'x',
              meta: {
                vis: {
                  component: 'Text'
                }
              }
            }
          }
        },
        {
          id: 'Bar',
          title: 'Bar',
          type: 'object',
          properties: {
            y: {
              type: 'string',
              title: 'Y',
              id: 'Y',
              meta: {
                vis: {
                  component: 'Text'
                }
              }
            }
          }
        }
      ]
    };
    const wrapper = mount(
      <Provider store={store}>
        <WrappedComponent
          schemaVis={{
            meta: {
              hasComponent: () => true
            },
            components: { Text },
            prefix: 'meta.vis',
            schema
          }}
          name="foobar"
        />
      </Provider>
    );
    const StyledComponent = wrapper.find(SingleSelect);
    const SelectComponent = StyledComponent.node.wrapped.wrappedInstance;
    wrapper.find(Input).simulate('change', {
      target: { value: 'Foo' }
    });
    expect(SelectComponent.state.selected).toEqual('Foo');
    wrapper.unmount();
    expect(StyledComponent.node.state.sheet.attached).toBeFalsy();
  });
});
