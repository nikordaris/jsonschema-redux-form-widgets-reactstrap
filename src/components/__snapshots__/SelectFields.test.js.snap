// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render SelectFields Should render SingleSelect 1`] = `
<div
  anyTouched={false}
  array={
    Object {
      "insert": [Function],
      "move": [Function],
      "pop": [Function],
      "push": [Function],
      "remove": [Function],
      "removeAll": [Function],
      "shift": [Function],
      "splice": [Function],
      "swap": [Function],
      "unshift": [Function],
    }
  }
  asyncValidate={[Function]}
  autofill={[Function]}
  blur={[Function]}
  className="form-group"
  clearAsyncError={[Function]}
  clearSubmit={[Function]}
  clearSubmitErrors={[Function]}
  destroy={[Function]}
  error={undefined}
  handleSubmit={[Function]}
  initialValues={undefined}
  initialize={[Function]}
  initialized={false}
  invalid={false}
  pure={true}
  reset={[Function]}
  submit={[Function]}
  submitSucceeded={false}
  touch={[Function]}
  triggerSubmit={undefined}
  type="select"
  valid={true}
  warning={undefined}
>
  <label
    className=""
    htmlFor="SingleSelect"
    id="SingleSelect-label"
  >
    SingleSelect
  </label>
  <select
    anyTouched={false}
    array={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "removeAll": [Function],
        "shift": [Function],
        "splice": [Function],
        "swap": [Function],
        "unshift": [Function],
      }
    }
    asyncValidate={[Function]}
    autofill={[Function]}
    blur={[Function]}
    className="form-control"
    clearAsyncError={[Function]}
    clearSubmit={[Function]}
    clearSubmitErrors={[Function]}
    destroy={[Function]}
    error={undefined}
    handleSubmit={[Function]}
    id="SingleSelect"
    initialValues={undefined}
    initialize={[Function]}
    initialized={false}
    invalid={false}
    name="SingleSelect"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    pure={true}
    reset={[Function]}
    submit={[Function]}
    submitSucceeded={false}
    touch={[Function]}
    triggerSubmit={undefined}
    valid={true}
    value=""
    warning={undefined}
  >
    <option
      disabled={true}
      value=""
    >
      Select 
      SingleSelect
    </option>
    <option
      value={2}
    >
      bar
    </option>
    <option
      value={1}
    >
      foo
    </option>
  </select>
</div>
`;

exports[`Render SelectFields Should render SingleSelect with only titles 1`] = `
<div
  anyTouched={false}
  array={
    Object {
      "insert": [Function],
      "move": [Function],
      "pop": [Function],
      "push": [Function],
      "remove": [Function],
      "removeAll": [Function],
      "shift": [Function],
      "splice": [Function],
      "swap": [Function],
      "unshift": [Function],
    }
  }
  asyncValidate={[Function]}
  autofill={[Function]}
  blur={[Function]}
  className="form-group"
  clearAsyncError={[Function]}
  clearSubmit={[Function]}
  clearSubmitErrors={[Function]}
  destroy={[Function]}
  error={undefined}
  handleSubmit={[Function]}
  initialValues={undefined}
  initialize={[Function]}
  initialized={false}
  invalid={false}
  pure={true}
  reset={[Function]}
  submit={[Function]}
  submitSucceeded={false}
  touch={[Function]}
  triggerSubmit={undefined}
  type="select"
  valid={true}
  warning={undefined}
>
  <label
    className=""
    htmlFor="SingleSelect"
    id="SingleSelect-label"
  >
    SingleSelect
  </label>
  <select
    anyTouched={false}
    array={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "removeAll": [Function],
        "shift": [Function],
        "splice": [Function],
        "swap": [Function],
        "unshift": [Function],
      }
    }
    asyncValidate={[Function]}
    autofill={[Function]}
    blur={[Function]}
    className="form-control"
    clearAsyncError={[Function]}
    clearSubmit={[Function]}
    clearSubmitErrors={[Function]}
    destroy={[Function]}
    error={undefined}
    handleSubmit={[Function]}
    id="SingleSelect"
    initialValues={undefined}
    initialize={[Function]}
    initialized={false}
    invalid={false}
    name="SingleSelect"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    pure={true}
    reset={[Function]}
    submit={[Function]}
    submitSucceeded={false}
    touch={[Function]}
    triggerSubmit={undefined}
    valid={true}
    value=""
    warning={undefined}
  >
    <option
      disabled={true}
      value=""
    >
      Select 
      SingleSelect
    </option>
    <option
      value="bar"
    >
      bar
    </option>
    <option
      value="foo"
    >
      foo
    </option>
  </select>
</div>
`;

exports[`Render SelectFields Should render SingleSelect with only values 1`] = `
<div
  anyTouched={false}
  array={
    Object {
      "insert": [Function],
      "move": [Function],
      "pop": [Function],
      "push": [Function],
      "remove": [Function],
      "removeAll": [Function],
      "shift": [Function],
      "splice": [Function],
      "swap": [Function],
      "unshift": [Function],
    }
  }
  asyncValidate={[Function]}
  autofill={[Function]}
  blur={[Function]}
  className="form-group"
  clearAsyncError={[Function]}
  clearSubmit={[Function]}
  clearSubmitErrors={[Function]}
  destroy={[Function]}
  error={undefined}
  handleSubmit={[Function]}
  initialValues={undefined}
  initialize={[Function]}
  initialized={false}
  invalid={false}
  pure={true}
  reset={[Function]}
  submit={[Function]}
  submitSucceeded={false}
  touch={[Function]}
  triggerSubmit={undefined}
  type="select"
  valid={true}
  warning={undefined}
>
  <label
    className=""
    htmlFor="SingleSelect"
    id="SingleSelect-label"
  >
    SingleSelect
  </label>
  <select
    anyTouched={false}
    array={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "removeAll": [Function],
        "shift": [Function],
        "splice": [Function],
        "swap": [Function],
        "unshift": [Function],
      }
    }
    asyncValidate={[Function]}
    autofill={[Function]}
    blur={[Function]}
    className="form-control"
    clearAsyncError={[Function]}
    clearSubmit={[Function]}
    clearSubmitErrors={[Function]}
    destroy={[Function]}
    error={undefined}
    handleSubmit={[Function]}
    id="SingleSelect"
    initialValues={undefined}
    initialize={[Function]}
    initialized={false}
    invalid={false}
    name="SingleSelect"
    onBlur={[Function]}
    onChange={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    pure={true}
    reset={[Function]}
    submit={[Function]}
    submitSucceeded={false}
    touch={[Function]}
    triggerSubmit={undefined}
    valid={true}
    value=""
    warning={undefined}
  >
    <option
      disabled={true}
      value=""
    >
      Select 
      SingleSelect
    </option>
    <option
      value="bar"
    >
      bar
    </option>
    <option
      value="foo"
    >
      foo
    </option>
  </select>
</div>
`;

exports[`Render SelectFields Should render object SingleSelect 1`] = `
<div
  className="card-0-10 card"
>
  <div
    className="header-0-11 card-header"
  >
    SingleSelect
  </div>
  <div
    className="card-block"
  >
    <select
      className="form-control"
      onChange={[Function]}
      value=""
    >
      <option
        disabled={true}
        value=""
      >
        Select 
        SingleSelect
      </option>
      <option
        value="Bar"
      >
        Bar
      </option>
      <option
        value="Foo"
      >
        Foo
      </option>
    </select>
  </div>
</div>
`;
